src/engine/Service.cpp

```diff
@@ -164,78 +166,109 @@ ProtoResult Service::computeResultImpl([[maybe_unused]] bool requestLaziness) {
   IdTableStatic<I> idTable = std::move(*idTablePtr).toStatic<I>();
   checkCancellation();
   std::vector<size_t> numLocalVocabPerColumn(idTable.numColumns());
 
-  size_t rowIdx = 0;
-  for (const auto& binding : bindings) {
-    idTable.emplace_back();
-    for (size_t colIdx = 0; colIdx < vars.size(); ++colIdx) {
-      TripleComponent tc = binding.contains(vars[colIdx])
-                               ? bindingToTripleComponent(binding[vars[colIdx]])
-                               : TripleComponent::UNDEF();
-
-      Id id = std::move(tc).toValueId(getIndex().getVocab(), *localVocab);
-      idTable(rowIdx, colIdx) = id;
-      if (id.getDatatype() == Datatype::LocalVocabIndex) {
-        ++numLocalVocabPerColumn[colIdx];
+  // Verifies that the result header contains the expected variables.
+  auto verifyVariables = [&](const nlohmann::json& part) {
+    auto vars = part["head"]["vars"].get<std::vector<std::string>>();
+    ad_utility::HashSet<Variable> responseVars;
+    for (const auto& v : vars) {
+      responseVars.emplace("?" + v);
+    }
+    ad_utility::HashSet<Variable> expectedVars(
+        parsedServiceClause_.visibleVariables_.begin(),
+        parsedServiceClause_.visibleVariables_.end());
+
+    if (responseVars != expectedVars) {
+      throw std::runtime_error(absl::StrCat(
+          "Header row of JSON result for SERVICE query is \"",
+          absl::StrCat("?", absl::StrJoin(vars, " ?")), "\", but expected \"",
+          absl::StrJoin(parsedServiceClause_.visibleVariables_, " ",
+                        Variable::AbslFormatter),
+          "\"."));
+    }
+  };
+
+  auto writeBindings = [&](const std::vector<nlohmann::json>& bindings,
+                           size_t& rowIdx) {
+    for (const auto& binding : bindings) {
+      idTable.emplace_back();
+      for (size_t colIdx = 0; colIdx < vars.size(); ++colIdx) {
+        TripleComponent tc =
+            binding.contains(vars[colIdx])
+                ? bindingToTripleComponent(binding[vars[colIdx]])
+                : TripleComponent::UNDEF();
+
+        Id id = std::move(tc).toValueId(getIndex().getVocab(), *localVocab);
+        idTable(rowIdx, colIdx) = id;
+        if (id.getDatatype() == Datatype::LocalVocabIndex) {
+          ++numLocalVocabPerColumn[colIdx];
+        }
       }
+      rowIdx++;
+      checkCancellation();
     }
-    rowIdx++;
-    checkCancellation();
+  };
+
+  size_t rowIdx = 0;
+  bool resultExists{false};
+  for (const nlohmann::json& part : body) {
+    if (part.contains("head")) {
+      AD_CORRECTNESS_CHECK(part["head"].contains("vars") &&
+                           part["head"]["vars"].is_array());
+      verifyVariables(part);
+    }
+    AD_CORRECTNESS_CHECK(part.contains("results") &&
+                         part["results"].contains("bindings") &&
+                         part["results"]["bindings"].is_array());
+    auto bindings =
+        part["results"]["bindings"].get<std::vector<nlohmann::json>>();
```

Comment:
I first thought `this is not a Correctness_check`, but now I am not so sure.
What does the lazy json parser do, if the array path is not found? does it throw? how does this interact with the SERVICE clause? There should be a unit test for this case. (maybe there already is).
But still you could add a comment, that this cann be a CHECK because the lazy parser already enforces these invariants.